File: NeetCode-150/Two Pointers/3 Sum/soln.java
class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        // JAI SHRI RAM 
        List<List<Integer>> res = new ArrayList<List<Integer>>(); 

        Arrays.sort(nums); 
        int n = nums.length; 

        for(int i = 0; i < n; i++){
            int l = i + 1; 
            int r = n - 1; 
            while(r > l){
                if(nums[l] + nums[r] == 0 - nums[i]){
                    
                    List<Integer> temp = new ArrayList<Integer>(); 
                    temp.add(nums[i]); 
                    temp.add(nums[l]); 
                    temp.add(nums[r]);
                    res.add(temp); 
                    while(l < n-1 && nums[l] == nums[l+1])
                        l++;
                    while(r > 0 && nums[r] == nums[r-1])
                        r--;
                    l++;
                    r--;
                    
                }
                else if(nums[l] + nums[r] > 0 - nums[i]){
                    r--;
                }
                else{
                    l++;
                }
            }
            while(i < n-1 && nums[i] == nums[i+1])
                i++;
            
        }
        return res;
    }
}

File: NeetCode-150/Two Pointers/3 Sum/soln.cpp
// Hari

// -1 -1 0 1 2 4
    vector<vector<int>> threeSum(vector<int>& nums) {
        // fast
        ios_base::sync_with_stdio(false);
        vector<vector<int>> res;
        
        sort(nums.begin(), nums.end());
        
        for(int i = 0; i<nums.size(); i++){
            int j = i+1, k = nums.size()-1;
            while(j<k) {
                if(nums[i]+nums[j]+nums[k] == 0) {
                    vector<int> temp;
                    temp.push_back(nums[i]);
                    temp.push_back(nums[j]);
                    temp.push_back(nums[k]);
                    res.push_back(temp);
                    
                    // skip duplicate j values
                    while(j<nums.size()-1 && nums[j] == nums[j+1]) j++;
                    // skip duplicate k values
                    while(k>0 && nums[k] == nums[k-1]) k--;
                    
                    // if no duplicates, then traverse both j and k
                    j++;
                    k--;
                    
                } else if (nums[i]+nums[j]+nums[k] > 0) {
                    k--;
                } else {
                    j++;
                }
            }
            
            // skip duplicate nums[i] because if a triplet was formed with this nums[i]
            // using it again will cause duplicate. If it didn't form triplet before, it will
            // not form triplet now also 
            while(i<nums.size()-1 && nums[i] == nums[i+1]) i++;
        }
        
        return res;
    }
